### PR TITLE
fix: DNS TTL not respected

### LINKED ISSUE #1425: DNS TTL Not Respected by celestia-core Leading to Potential Future Sync Issues
**Description:**

We identified a potential issue where changes to the DNS entries of nodes in celestia-core are not respected, similar to the issue observed in celestia-node. Although we have not yet encountered sync issues due to this behavior, it could pose a problem in the future. It appears that nodes resolve DNS entries only once at startup and continue using the same IP address indefinitely, ignoring DNS TTL.

**Steps to Reproduce:**
1. Change the DNS entries for nodes.
2. Observe that nodes continue to use the old IP address without re-resolving the DNS entries.

**Suspected Cause:**
Nodes resolve DNS entries only once at startup and continue using the same IP address without respecting the TTL. This could affect:
- Nodes connecting to other nodes within the network.

**Relevant Code:**
- DNS resolution implementation in celestia-core: https://github.com/celestiaorg/celestia-core/blob/c2e6331b8f0c920ff68f27e8f89f97a18832e273/p2p/peer.go#L140

**Potential Fix:**
1. Periodically re-resolve DNS entries based on the TTL.
2. Update active connections if the resolved IP address changes.

**Repositories Potentially Needing Changes:**
- [`celestiaorg/celestia-core`](https://github.com/celestiaorg/celestia-core)
- [`cometbft/cometbft`](https://github.com/cometbft/cometbft) (upstream implementation of Tendermint)

**Impact:**
Not respecting DNS TTL can lead to potential connectivity and sync issues, affecting network reliability in the future.

**Request for Assistance:**
1. Implement periodic DNS resolution based on TTL.
2. Test changes to ensure nodes dynamically update connections based on DNS updates.

### PR DESCRIPTION
**Pull request by ninabarbakadze**
## Description

Fixes #1425 

## Testing

Testing DNS change is not really possible in unit tests so @smuu agreed to help me test manually

### CODE REVIEW

File: p2p/peer.go [hunk 1]
```diff
@@ -138,10 +140,6 @@ func (pc peerConn) ID() ID {
 
 // Return the IP from the connection RemoteAddr
 func (pc peerConn) RemoteIP() net.IP {
-	if pc.ip != nil {
-		return pc.ip
-	}
```
Comment (cmwaters) [inline p2p/peer.go:143]: I think the idea of caching the result so we don't need to resolve the IP every time this method is called still makes sense. Perhaps what you need to do when checking `IPHasChanged`, is to set the p.ip to nil before calling `RemoteIP`

File: p2p/peer.go [hunk 2]
```diff
@@ -300,6 +298,16 @@ func (p *peer) IsPersistent() bool {
 	return p.peerConn.persistent
 }
 
+// IPHasChanged returns true and the new IP if the peer's IP has changed.
+func (p *peer) IPHasChanged() bool {
```
Comment (cmwaters) [inline p2p/peer.go:302]:
```suggestion
func (p *peer) HasIPChanged() bool {
```
nit: but this sounds better

**Review by cmwaters — CHANGES_REQUESTED** [review]

File: p2p/peer.go (hunk 1 above)
Comment (ninabarbakadze) [reply p2p/peer.go]: updated

**Review by cmwaters — CHANGES_REQUESTED** [review]
This all looks good. The only hesitation I have is that `reconnectPeer` is relentless. It's designed for persistent peers, not for normal peers that you want to connect to. Thus, afaik, it ignores errors and continuously redials.

When trying to reconnect, if this fails we should give up trying to connect the peer. I would suggest we add some additional input to the `reconnectToPeer` method which will distinguish between retry forever and retry once and if failed log an error

**Comment by ninabarbakadze** [comment]
> This all looks good. The only hesitation I have is that `reconnectPeer` is relentless. It's designed for persistent peers, not for normal peers that you want to connect to. Thus, afaik, it ignores errors and continuously redials.
> 
> When trying to reconnect, if this fails we should give up trying to connect the peer. I would suggest we add some additional input to the `reconnectToPeer` method which will distinguish between retry forever and retry once and if failed log an error

For context here the peer won't try to reconnect forever. It first tries to connect 20 times with 5 second intervals(20*5=100s), then another 10 times with exponential backoff which ends in 16h(3**10 = 16hrs). it will give up after that so no change needed.

**Review by cmwaters — APPROVED** [review]
👍  Thanks @ninabarbakadze

**Review by evan-forbes — APPROVED** [review]

**Review by ninabarbakadze — COMMENTED** [review]
should i merge this? @cmwaters @evan-forbes 

I don't think dev ops was able to test this yet

**Comment by evan-forbes** [comment]
yeah, we should try it out

**Pull request merged**